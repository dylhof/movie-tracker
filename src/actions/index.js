export const storeMovies = (movies) => ({
  type: 'STORE_MOVIES',
  movies
})

export const setCurrentUser = (name, id) => ({
  type: 'SET_USER',
  name,
  id
})

export const logoutCurrentUser = () => ({
  type: 'LOGOUT_USER'
})

export const addFavorite = (id) => ({
  type: 'ADD_FAVORITE',
  id
})

export const addAllUserFavorites = (favorites) => ({
  type: 'ADD_ALL_USER_FAVORITES',
  favorites
})

export const deleteFavorite = (id) => ({
  type: 'DELETE_FAVORITE',
  id
})

export const clearFavorites = () => ({
  type: 'CLEAR_FAVORITES',
})

export const setLoading = (bool) => ({
  type: 'SET_LOADING',
  bool
})