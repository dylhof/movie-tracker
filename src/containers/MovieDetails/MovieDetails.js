import React, { Component } from 'react';
import { connect } from 'react-redux';

export class MovieDetails extends Component{
  
  render() {
    const {title, release_date, overview, poster_path} = this.props
    const poster = `https://image.tmdb.org/t/p/w200/${poster_path}`
    const alt = `${title} poster`
    return (
      <div>
        <h3>{title}</h3>
        <img src={poster} alt={alt} />
        <p>{release_date}</p>
        <p>{overview}</p>
      </div>
    )
  }
}

export const mapStateToProps = (state) => ({
  currentUser: state.currentUser,
})

export default connect(mapStateToProps)(MovieDetails)