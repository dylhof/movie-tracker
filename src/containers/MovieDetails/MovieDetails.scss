.movie-details-div {
  background: #0D0707;
  box-shadow: 5px 5px 7px #2b0c0e;
  width: 75%;
  margin: 50px auto;
  border-radius: 10px;
  display: flex;

  p, h2 {
    color: $font-color;
    margin-right: 20px;
  }

  h2 {
    height: 55px;
    margin-bottom: 5px
  }

  img {
    border: 4px solid $font-color;
    border-radius: 3px;
    margin: 20px;
  }

  .movie-details-title-div {
    display: flex;
    justify-content: space-between;
  }

  button{
    margin: 20px 20px 25px 0px;
    
    &:focus {
      outline: 0;
    }
    &:hover {
      color: $font-color;
    }
  }

  .overview {
    height: 150px;
    overflow: scroll;
  }

}