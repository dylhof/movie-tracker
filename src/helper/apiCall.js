export const fetchData = async (url) => {
  const response = await fetch(url)
  if(response.ok) {
    return response.json() 
  } else {
    throw new Error(`Error fetching, code: ${response.status}`)
  }
}

export const fetchPost = async (url, options) => {
  const response = await fetch(url, options)
  if (response.ok) {
    return response.json()
  } else {
    throw new Error(response.status)
  }
}