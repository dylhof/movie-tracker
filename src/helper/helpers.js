import * as API from './apiCall';

export const tryUnfavorite = async (id, userID) => {

  const url = `http://localhost:3000/api/users/${userID}/favorites/${id}`
  await API.fetchPost(url,
    {
      method: 'DELETE',
      body: JSON.stringify({
        user_id: userID, movie_id: id
      }),
      headers: {
        'Content-Type': 'application/json'
      }
    })
}

export const tryFavorite = async (title, id, userID, poster_path, release_date, vote_average, overview) => {

  await API.fetchPost('http://localhost:3000/api/users/favorites/new',
    {
      method: 'POST',
      body: JSON.stringify({
        movie_id: id, user_id: userID, title, poster_path,
        release_date, vote_average, overview
      }),
      headers: {
        'Content-Type': 'application/json'
      }
    })
}

