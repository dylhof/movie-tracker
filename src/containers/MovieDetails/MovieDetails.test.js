import React from 'react';
import { shallow } from 'enzyme';
import { MovieDetails, mapStateToProps, mapDispatchToProps } from './MovieDetails'
import * as helper from '../../helper/helpers'
import * as actions from '../../actions/index'

describe('MovieDetails', () => {
  let wrapper
  beforeEach(() => {
    let currentUser = { name: 'Matt', userID: 1 }
    let mockDispatchDeleteFavorite = jest.fn()
    let mockDispatchAddFavorite = jest.fn()
    let mockDispatchSetError = jest.fn()
    let mockFavorites = [1, 2, 3]
    wrapper = shallow(
      <MovieDetails
        currentUser={currentUser}
        dispatchDeleteFavorite={mockDispatchDeleteFavorite}
        dispatchAddFavorite={mockDispatchAddFavorite}
        dispatchSetError={mockDispatchSetError}
        favorites={mockFavorites}
      />
    )
  })

  it('should match the snapshot', () => {
    //expectation
    expect(wrapper).toMatchSnapshot()
  })

  describe('handleFavoriteClick', () => {

    it('should call tryUnfavorite if currentUser and already favorite', () => {
      //setup
      const mockEvent = { target: { value: 'true' } }
      helper.tryUnfavorite = jest.fn()
      //execution
      wrapper.instance().handleFavoriteClick(mockEvent)
      //expectation
      expect(wrapper.instance().props.dispatchDeleteFavorite).toHaveBeenCalled()
      expect(helper.tryUnfavorite).toHaveBeenCalled()
    })

    it('should call tryFavorite if currentUser and not already favorite', () => {
      //setup 
      const mockEvent = { target: { value: 'false' } }
      helper.tryFavorite = jest.fn()
      //execution
      wrapper.instance().handleFavoriteClick(mockEvent)
      //expectation
      expect(wrapper.instance().props.dispatchAddFavorite).toHaveBeenCalled()
      expect(helper.tryFavorite).toHaveBeenCalled()
    })

    it('should set state of isUser if there is no current user', () => {
      //setup
      const mockEvent = { target: { value: 'false' } }
      let mockFavorites = [1, 2, 3]
      let currentUser = null
      wrapper = shallow(
        <MovieDetails
          currentUser={currentUser}
          favorites={mockFavorites}
        />)
      //execution
      wrapper.instance().handleFavoriteClick(mockEvent)
      //expectation
      expect(wrapper.state('isUser')).toEqual('Please login or create an acount to favorite your movies!')
    })
  })

  describe('mapStateToProps', () => {

    it('should return an object with a current user and an array of favorites', () => {
      //setup
      const mockState = {
        currentUser: [{ title: 'Aquaman' }],
        favorites: [{}],
        error: 'Error'
      }
      const expected = {
        currentUser: [{ title: 'Aquaman' }],
        favorites: [{}]
      }
      //execution
      const mappedProps = mapStateToProps(mockState)
      //expectation
      expect(mappedProps).toEqual(expected)
    })
  })

  describe('mapDispatchToProps', () => {

    it('should call dispatch when using dispatchAddFavorite from MDTP', () => {
      //setup
      const mockDispatch = jest.fn()
      const actionToDispatch = actions.addFavorite(1)
      //execution
      const mappedProps = mapDispatchToProps(mockDispatch)
      mappedProps.dispatchAddFavorite(1)
      //expectation
      expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch)
    })

    it('should call dispatch when using dispatchDeleteFavorite from MDTP', () => {
      //setup
      const mockDispatch = jest.fn()
      const actionToDispatch = actions.deleteFavorite(1)
      //execution
      const mappedProps = mapDispatchToProps(mockDispatch)
      mappedProps.dispatchDeleteFavorite(1)
      //expectation
      expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch)
    })
  })

})