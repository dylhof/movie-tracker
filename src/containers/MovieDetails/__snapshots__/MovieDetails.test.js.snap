// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MovieDetails should match the snapshot 1`] = `
<div
  className="movie-details-div"
>
  <img
    alt="undefined poster"
    src="https://image.tmdb.org/t/p/w500/undefined"
  />
  <div>
    <div
      className="movie-details-title-div"
    >
      <h2 />
      <button
        className="favorite-btn "
        onClick={[Function]}
        value={false}
      >
        <i
          className="fas fa-star"
        />
      </button>
    </div>
    <p
      className="overview"
    />
    <p>
      Original Release Date: 
    </p>
    <p>
      The Movie DataBase Vote Average: 
    </p>
  </div>
</div>
`;
